plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts

```diff
@@ -231,6 +232,23 @@ export class AwsLambdaInstrumentation extends InstrumentationBase {
     this._tracerProvider = tracerProvider;
   }
 
+  private _getBasicTracerProvider() {
+    if(!this._tracerProvider) {
+      diag.error('Spans are not exported for the lambda function as the tracerProvider is undefined.');
+      return undefined;
+    }
+
+    let realTracerProvider = this._tracerProvider;
+
+    if (this._tracerProvider instanceof ProxyTracerProvider) {
```

Comment:
I'm not sure if the `instanceof` checks work if there are more instances of the `@opentelemetry/api` package installed.

I have following setup in mind where more instances (of the same/compatible version) are installed:
* a layer holding the AWS part of the instrumentation for AWS services
* user handler has it's own  `@opentelemetry/api` in it's dependencies as it also calls OTel APIs (and/or adds more instrumentations)

Reply:
@Flarna What do you suggest? For me, I believe @willarmiros may be on the same page is that I believe the happy path right now is broken and get that fix and figure out the proper fix later. This will be an incremental, low effort solution for now.

Reply:
What about `this._tracerProvider.getDelegate instanceof 'function'`?

Reply:
I think you meant `typeof this._tracerProvider.getDelegate === 'function'` @dyladan . I'll let y'all hash it out before I make any other changes. I'm going for the low effort, low risk solution. If desire any more, I may not have the time to help.

Reply:
I was also thinking about `typeof` function checks.

How is this instrumentation used in general? e.g. how is it instantiated?
If there is no risk that more `@opentelemetry/api` instances are present or if it is ensured from the initialization flow that the `@opentelemetry/api` instance referenced by this instrumentation is the same as that one used to create the global `TracerProvider` it's fine to use `instanceof`.

Reply:
> If there is no risk that more @opentelemetry/api instances are present or if it is ensured from the initialization flow that the @opentelemetry/api instance referenced by this instrumentation is the same as that one used to create the global TracerProvider it's fine to use instanceof

I think since Lambda sandboxes are just that, sandboxes, we can make no guarantee of what users will have in them. So it's probably safer to go with the `typeof this._tracerProvider.getDelegate === 'function'` check that we all seem to be in agreement on. @longility could you update the PR? I'm comfortable with this solution for now since it's a pretty critical, blocking bug.

Reply:
Should we do function check for only `getDelegate` or should we do it for `getActiveSpanProcessor` and `forceFlush` as well?

Reply:
I would check all 3 since those methods aren't required by the API a third party sdk which doesn't have those functions could theoretically be used.

Reply:
I updated based on what is suggested, and I'm getting test failures to where I need more help.

```
25) lambda handler
       hooks
         responseHook
           sync - error:
     TypeError: Cannot read property 'attributes' of undefined
      at Context.<anonymous> (test/integrations/lambda-handler.test.ts:18:48)
      at processImmediate (internal/timers.js:464:21)
      at process.callbackTrampoline (internal/async_hooks.js:131:17)

  26) lambda handler
       hooks
         responseHook
           sync - error with callback:
     TypeError: Cannot read property '_spanProcessors' of undefined
      at forceFlush (node_modules/@opentelemetry/tracing/src/MultiSpanProcessor.ts:33:38)
      at AwsLambdaInstrumentation._endSpan (src/instrumentation.ts:66:4)
      at /Users/longility/Repos/longility-opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:65:33
      at Object.safeExecuteInTheMiddle (node_modules/@opentelemetry/instrumentation/src/utils.ts:36:5)
      at /Users/longility/Repos/longility-opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts:64:15
      at async_hooks.js:312:14
      at AsyncResource.runInAsyncScope (async_hooks.js:197:9)
      at AsyncLocalStorage.run (async_hooks.js:310:35)
      at AsyncLocalStorageContextManager.with (node_modules/@opentelemetry/context-async-hooks/src/AsyncLocalStorageContextManager.ts:40:36)
      at ContextAPI.with (node_modules/@opentelemetry/api/src/api/context.ts:77:42)
      at Object.patchedHandler [as callbackerror] (src/instrumentation.ts:46:46)
      at /Users/longility/Repos/longility-opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-aws-lambda/test/integrations/lambda-handler.test.ts:18:48
      at new Promise (<anonymous>)
      at Context.<anonymous> (test/integrations/lambda-handler.test.ts:18:48)
      at processImmediate (internal/timers.js:464:21)
      at process.callbackTrampoline (internal/async_hooks.js:131:17)
```

Reply:
Sorry for the late reply @longility, I've only just got around to debugging. It turns out we're just not passing in the `thisArg` when calling forceFlush. I'll make some suggestions on the PR to address.

Also, if you don't have the time to address this, I can roll these changes into #660 and give you a co-author :) up to you

Reply:
Thanks @willarmiros I committed your suggestions and that worked! I learned about suggested changes. Thanks for that. That is nice.

Also, can you elaborate on your comment as I would like to learn. The changes was not enough to understand why that worked and my changes does not.
> It turns out we're just not passing in the thisArg when calling forceFlush. I'll make some suggestions on the PR to address.

Reply:
> Also, can you elaborate on your comment as I would like to learn. The changes was not enough to understand why that worked and my changes does not.

In `strict mode` JS, when you call a function of an object, it implicitly passes the object to the function which can be referenced as `this`. For instance, if you call `foo.bar()`, you can reference `foo` within the `bar` function via `this`.

In the original change, since we were calling `forceFlush` without a reference to its parent object, `this` was undefined within the `forceFlush` function, causing an error [here](https://github.com/open-telemetry/opentelemetry-js/blob/bebd82a7f12927bb8a486a252bb7c1cc12619294/packages/opentelemetry-sdk-trace-base/src/MultiSpanProcessor.ts#L33).

So my change allows the `this` reference to be defined correctly. Read more: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this

Reply:
Makes sense, thanks! Alternatively, it has been a while since I had to care about `this`, but if I recall, we could explicitly `bind` to its original objects as well. Thanks for taking your time to explain a gotcha that I was unaware of @willarmiros !